### PR TITLE
Feature/changelog

### LINKED ISSUE #54: Add a CHANGELOG.md file
There are some solutions for automating the changelog creation.

Also, it shouldn't be a priority until we hit v1.0 since changes will occur rapidly now in alpha/beta versions

### PR DESCRIPTION
**Pull request by JayaKrishnaNamburu**
Fixes #54 

Hi @alexnm I gave a try for both [semantic-release](https://github.com/semantic-release/semantic-release) and [bump](https://github.com/fabiospampinato/bump) well semantic-release looked very matured but after trying out to implement it on an example repo, I realized ``bump`` is something that we need here. Here are the observations for both

Pros for ``semantic-release``
- Can be integrated with Travis so releases happen automatically depending on the commit messages.
- A release is published in both GitHub and npm as well.

Cons for this case as the whole eco-system is in beta/alpha

- By default ``semantic-release`` starts its versioning with v1.0
- Maintaining pre-release, beta is a little tricky. 
- The configuration is a little [tricky](https://github.com/semantic-release/npm/issues/151) 

Why Bump?

- Generates changelog and updates it whenever we bump a ``major/minor/path`` is added.
- Releasing a version is totally controlled by the developer using ``bump`` command (That is where semantic-release helps to reduce human errors but it helps in production releases the most).
- Pushes to GitHub and drafts a release with selected fields and updates the changelog using ``bump.json`` file. 

~~~
bump changelog // generates changelog
bump // gives a list of options to raise the bump
bump release // releases/drafts to GitHub
~~~

Screenshot 2019-03-15 at 11 46 29 AM
Screenshot 2019-03-15 at 11 46 48 AM

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+- Merge pull request #69 from paulbrie/master
+- fix: tests
+- Merge pull request #71 from teleporthq/cleanup-builder-and-chunk
+- test(component generators): tests for all style flavors
+- Merge branch 'master' of https://github.com/mihaiserban/teleport-code-generators into ISSUE-52
+- Merge pull request #68 from mihaiserban/ISSUE-52
+- refactor(key uniqueness): UIDL structure update
+- test: temporary threshold for performance tests
+
+### Version 0.3.1
```
Comment (alexnm) [inline CHANGELOG.md]: there seems to be a small mixup here with 0.3.1 shown twice and once empty. this file is autogenerated when running bump?

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
+### Version 0.4.1
+- Merge pull request #74 from teleporthq/hotfix/0.4
```
Comment (alexnm) [inline CHANGELOG.md]: Is there a way to not show the merges in here and actually show the relevant commit messages?

File: CHANGELOG.md (hunk 1 above)
Comment (JayaKrishnaNamburu) [reply CHANGELOG.md]: Yes, the file is auto-generated from when we run ``bump changelog``

File: CHANGELOG.md (hunk 2 above)
Comment (JayaKrishnaNamburu) [reply CHANGELOG.md]: Yeah, I will look into that 👍

**Comment by JayaKrishnaNamburu** [comment]
And @alexnm does the Readme.md file needs any changes because the ``install`` script which is given as an example is not the same as the that is being imported, I tried in an example and when i changed the name back the code is being generated 🤔

File: CHANGELOG.md (hunk 1 above)
Comment (alexnm) [reply CHANGELOG.md]: if we manually fix this will it be overwritten when we run `bump changelog` again?

Comment (JayaKrishnaNamburu) [reply CHANGELOG.md]: I just tested it out by manually correcting in my local, it's not overwriting the changelog, instead it's just writing the new changes :)

File: bump.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (alexnm) [inline bump.json]: you could also add `bump` as a `devDependency` so se can use it as an `npm` task and then no one needs to install it as a global package

**Comment by alexnm** [comment]
@JayaKrishnaNamburu good catch with the readme file, it's supposed to be imported from `@teleporthq/teleport-code-generators`. I will fix the readme

File: bump.json (hunk 3 above)
Comment (JayaKrishnaNamburu) [reply bump.json]: I remember adding it, don't know how it slipped, will do that now 👍

**Comment by JayaKrishnaNamburu** [comment]
> @JayaKrishnaNamburu good catch with the readme file, it's supposed to be imported from `@teleporthq/teleport-code-generators`. I will fix the readme

It should be updated in few other places as well -- > https://github.com/teleporthq/teleport-code-generators#react 😄

**Comment by alexnm** [comment]
already updated in all the places in the readme

File: CHANGELOG.md (hunk 2 above)
Comment (JayaKrishnaNamburu) [reply CHANGELOG.md]: Hi @alexnm  currently ``bump`` doesn't support the omission of merge commits out of the box but I created a [PR](https://github.com/fabiospampinato/bump/pull/5) on their Repo, let's see 😄 

Here is how the changelog will look if that is implemented. 

~~~
### Version 0.4.1
- refactor(resolver): uniform implementation of util functions
- refactor(deprecation): deprecate react-all generator
- fix code samples
- Update README.md
- refactor(dependencies): move extract dependencies inside assembly line
- fix(vue output): remove style tag if it is not needed
- refactor(plugins): using a generic traversal method
- refactor(style plugins): handling dynamic values in nested styles (origin/master, origin/HEAD, master)
- chore(bump.json): Added bump.json for generating changelogs and releases
~~~

Comment (alexnm) [reply CHANGELOG.md]: that looks awesome, I think we can integrate it like this for now, if we can get this feature into `bump` at a certain point in the future

**Comment by JayaKrishnaNamburu** [comment]
Hi @alexnm I think there is one more place where ``Readme`` needs a change --> https://github.com/teleporthq/teleport-code-generators#uidl-validators

File: bump.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        "tag": {
+            "enabled": true,
+            "name": "v[version]"
+        },
+        "github": {
+            "enabled": true,
+            "open": true,
+            "draft": true,
+            "prerelease": true,
+            "token": ""
```
Comment (JayaKrishnaNamburu) [inline bump.json]: A ``GIT_HUB`` token should be updated here in order to push the newly tagged version and drat a release :)

**Comment by alexnm** [comment]
@JayaKrishnaNamburu we're going to discuss these days about switching to `lerna` or `yarn workspaces` so we are postponing this a bit until that is cleared. See #30 for reference

**Comment by Utwo** [comment]
PR looks good! We only need to specify non-default values in bump.json and to update bump version in package.json. Something like this:

```{
  "version": {
    "initial": "0.4.1",
  },
  "changelog": {
    "create": true,
    "open": false
  },
  "release": {
    "enabled": true,
    "github": {
      "enabled": true,
      "open": true,
      "draft": false,
      "prerelease": true,
      "token": ""
    }
  }
}
```

**Comment by Utwo** [comment]
Any update on this?

**Comment by JayaKrishnaNamburu** [comment]
Hi, @alexnm updated and squashed the commits, Please review.

**Comment by alexnm** [comment]
I will ask @Utwo to take a look as I'm caught up with the Vue generators

**Comment by Utwo** [comment]
Thank you! :+1:

**Pull request merged**